wall_e/extensions/sfu.py

```diff
@@ -502,6 +506,104 @@ async def outline(self, ctx, *course):
+            self.logger.debug(f"[SFU courses()] url for get constructed: {url}")
+
+            res = await self.req.get(url)
+            if res.status == 200:
+                self.logger.debug(f"[SFU courses()] get request for {department} successful, parsing data")
+                res_json = await res.json()
+
+                # parse data for displaying, sorting, and filtering
+                for course in res_json:
+                    course["text"] = f"{department}{course['text']}"
+
+                    # we assume all courses have 3 digits
+                    if len(course["value"]) == 4:
+                        course["value"] = course["value"][:3]
+
+                    course["value"] = int(course["value"])
+                    if level == 0 or (course["value"]//100 == level//100):
+                        courses.append(course)
+            else:
+                self.logger.debug(f"[SFU courses()] get request for {department} resulted in {res.status}")
+
+        if len(courses) == 0:
+            self.logger.debug("[SFU courses()] resulted in no content")
+            e_obj = await embed(
+                self.logger, interaction=interaction, title="SFU Courses Error",
+                description=(
+                    f"Couldn't find anything for `department: {', '.join(departments)}"
+                    f", level: {level if level != 0 else 'all'}, term: {term}, year: {year}`"
+                    f"\n Maybe no courses are being offered at that time."
+                ),
+                colour=WallEColour.ERROR,
+            )
+            if e_obj:
+                msg = await interaction.followup.send(embed=e_obj)
+                await asyncio.sleep(10)
+                await msg.delete()
+            return
+
+        if len(departments) > 1:
+            courses = sorted(courses, key=lambda k: k["value"])
+
+        self.logger.debug("[SFU courses()] parsing data from get request")
+        content_to_embed = []
+        number_of_courses_per_page = 20
+        number_of_courses = 0
+        total_course = 0
+        content = ""
+
+        for i, course in enumerate(courses):
+            course_title = course["title"]
+            course_number = course["text"]
+            content += f"\n{course_number} - {course_title}"
+
+            total_course += 1
+            number_of_courses += 1
+            if total_course > 0 and (number_of_courses % number_of_courses_per_page == 0 or i == len(courses) - 1):
+                number_of_courses = 0
+                content_to_embed.append(
+                    [["Code - Title", content]]
+                )
+                content = ""
+
+        title = (f"{', '.join(departments)} {f'{(level//100)*100} level -' if level != 0 else '-'}"
+                 f" {f'{term.title()}' if term != 'registration' and term != 'current' else f'{term} term'}"
+                 f" {f'{year.title()}' if year != 'registration' and year != 'current' else f'{year} year'}"
```

Comment:
Oh! And instead of displaying "registration" on the discord embed, could you display "next term"? I think it would be easier for users to understand (I wasn't sure what registration was referring to until I read the sfu api)

Reply:
one of the reason i did this was because if the user were to enter for example: `/courses year:2021`, the sfu api would respond with the registration term, which at the time of writing is fall (current and registration seem to be the same at the moment), but for the year 2021, therefore purely next term would be incorrect

i could change it so that if the term and registration is default, it would display "next term", and otherwise, the same as before? or would "next term 2021" be better? 

also similarly, if the user were to enter `/courses term:summer`, it would be the registration year but for the summer semester

the other option would be to only allow change in input to year or term only if both are changed

Reply:
Ahh, I see your intentions.

> i could change it so that if the term and registration is default, it would display "next term", and otherwise, the same as before? or would "next term 2021" be better?

I think either are fine, we just want to hide the implementation detail of "registration" from the user. I would probably prefer just "next term", as it's clear what year the user input. If the user only provides `term:summer`, then maybe "next summer" would be a reasonable solution?

Reply:
One problem with "next summer" is that it could be misleading, such as with right now, SFU seems to give back the courses for this year's summer. And in the case that it would give the correct results, I do not believe SFU gives courses for classes 2 semesters ahead, so it would only display no results.

Therefore, I believe that the best option would be to have both arguments be specified in order to work.

Reply:
ahh, darn. But good to know! I agree, that'd definitely be misleading.

> to have both arguments be specified in order to work

This is a good idea! Passing just "2022" without a semester doesn't make too much sense from the user's point of view. And if someone wants "next summer", it's easy enough to know that next semester is 2025.

Reply:
https://github.com/CSSS/wall_e/pull/666/commits/5f54299272d64864628b480c63f3459a4236650e